pkg/apis/certificates/v1beta1/defaults.go

```diff
@@ -17,15 +17,113 @@ limitations under the License.
 	return RegisterDefaults(scheme)
 }
+
 func SetDefaults_CertificateSigningRequestSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) {
 	if obj.Usages == nil {
 		obj.Usages = []certificatesv1beta1.KeyUsage{certificatesv1beta1.UsageDigitalSignature, certificatesv1beta1.UsageKeyEncipherment}
 	}
+
+	if obj.SignerName == nil {
+		signerName := DefaultSignerNameFromSpec(obj)
+		obj.SignerName = &signerName
+	}
+}
+
+// DefaultSignerNameFromSpec will determine the signerName that should be set
+// by attempting to inspect the 'request' content and the spec options.
+func DefaultSignerNameFromSpec(obj *certificatesv1beta1.CertificateSigningRequestSpec) string {
+	csr, err := ParseCSR(obj.Request)
+	switch {
+	case err != nil:
+		// Set the signerName to 'legacy-unknown' as the CSR could not be
+		// recognised.
+		return certificatesv1beta1.LegacyUnknownSignerName
+	case IsKubeletClientCSR(csr, obj.Usages):
+		return certificatesv1beta1.KubeAPIServerClientKubeletSignerName
+	case IsKubeletServingCSR(csr, obj.Usages):
+		return certificatesv1beta1.KubeletServingSignerName
+	default:
+		return certificatesv1beta1.LegacyUnknownSignerName
+	}
+}
+
+func IsKubeletServingCSR(req *x509.CertificateRequest, usages []certificatesv1beta1.KeyUsage) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, req.Subject.Organization) {
+		return false
+	}
+
+	// at least one of dnsNames or ipAddresses must be specified
+	if len(req.DNSNames) == 0 && len(req.IPAddresses) == 0 {
+		return false
+	}
+
+	if len(req.EmailAddresses) > 0 || len(req.URIs) > 0 {
+		return false
+	}
+
+	requiredUsages := []certificatesv1beta1.KeyUsage{
+		certificatesv1beta1.UsageDigitalSignature,
+		certificatesv1beta1.UsageKeyEncipherment,
+		certificatesv1beta1.UsageServerAuth,
+	}
+	if !equalUnsorted(requiredUsages, usages) {
+		return false
+	}
+
+	if !strings.HasPrefix(req.Subject.CommonName, "system:node:") {
+		return false
+	}
+
+	return true
+}
+
+func IsKubeletClientCSR(req *x509.CertificateRequest, usages []certificatesv1beta1.KeyUsage) bool {
+	if !reflect.DeepEqual([]string{"system:nodes"}, req.Subject.Organization) {
+		return false
+	}
+
+	if len(req.DNSNames) > 0 || len(req.EmailAddresses) > 0 || len(req.IPAddresses) > 0 || len(req.URIs) > 0 {
+		return false
+	}
+
+	if !strings.HasPrefix(req.Subject.CommonName, "system:node:") {
+		return false
+	}
+
+	requiredUsages := []certificatesv1beta1.KeyUsage{
+		certificatesv1beta1.UsageDigitalSignature,
+		certificatesv1beta1.UsageKeyEncipherment,
+		certificatesv1beta1.UsageClientAuth,
+	}
+	if !equalUnsorted(requiredUsages, usages) {
+		return false
+	}
+
+	return true
+}
+
+// equalUnsorted compares two []string for equality of contents regardless of
+// the order of the elements
+func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
+	l := sets.NewString()
+	for _, s := range left {
+		l.Insert(string(s))
+	}
+	r := sets.NewString()
+	for _, s := range right {
```

Comment:
It seems there is no need to create the second Set.
Checking sizes of the slices and whether each of the second slice is contained in the first Set should suffice.

Reply:
that would return true for `[a, b, c]` vs `[a, a, b]`

Reply:
The following would achieve same semantics with one map.
```
@@ -117,13 +116,23 @@ func IsKubeletClientCSR(req *x509.CertificateRequest, usages []certificatesv1bet
 // equalUnsorted compares two []string for equality of contents regardless of
 // the order of the elements
 func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
-       l := sets.NewString()
+       var l map[string]int
+       l = make(map[string]int)
        for _, s := range left {
-               l.Insert(string(s))
+               k := string(s)
+               cnt := l[k]
+               l[k] = cnt + 1
        }
-       r := sets.NewString()
        for _, s := range right {
-               r.Insert(string(s))
-       }
-       return l.Equal(r)
+               k := string(s)
+               cnt, ok := l[k]
+               if !ok {
+                       return false
+               }
+               cnt--
+               if cnt == 0 {
+                       delete(l, k)
+               }
+       }
+       return len(l) == 0
 }
```

Reply:
if we really want to be cheap:
```go
func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
	if len(left) != len(right) {
		return false
	}
	l := make(map[string]int, len(left))
	for _, s := range left {
		k := string(s)
		l[k] = l[k] + 1
	}
	for _, s := range right {
		k := string(s)
		c := l[k]
		if c == 0 {
			return false
		}
		l[k] = (c - 1)
	}
	return true
}
```
but we do use the sets API all over, I'm not sure if this change is desirable.

Reply:
```
	if len(left) != len(right) {
		return false
	}
```
We shouldn't use the above check since there may be duplicate in either side.
```
		c := l[k]
		if c == 0 {
```
The decrement should be done before checking c against 0.

I prefer the formation I presented above.
Set is not expressive in this use case. We don't have to use it.

Reply:
> We shouldn't use the above check since there may be duplicate in either side.

If we're not checking for _identical_ counts in each side, why even count in the map ...?

> The decrement should be done before checking c against 0.

No, if we decrement to zero then we've had an exact match. If we find an entry with zero on the next check we did not. I did test this code 🙃

Reply:
Thanks @BenTheElder for looking at my code.

Before we dig deeper into the code, can we get agreement on whether the following are considered equal:

[a, b] vs [a, a, b]

Along the current code, they're equal.
That was why I said we shouldn't check the lengths of two sides.

Reply:
@BenTheElder was right in that the value of the map shouldn't be the number of times we see the same key from the left.
Here is revised code which keeps the current semantics and only uses one map:
```
func equalUnsorted(left, right []certificatesv1beta1.KeyUsage) bool {
	var l map[string]int
	l = make(map[string]int)
	for _, s := range left {
		k := string(s)
		l[k] = 1
	}
	count := 0
	for _, s := range right {
		k := string(s)
		val, ok := l[k]
		if !ok {
			// not equal if we haven't seen this key from the left
			return false
		}
		if val == 1 {
			count++
			// we don't want to delete the key since there may be more occurrence of this key in right
			l[k] = -1
		}
	}
	// if we have seen all the keys from the right, they're equal
	return len(l) == count
}
```
I can refine the comment in the code if there is question on how it works or whether it is correct.

Thanks

Reply:
We can keep this on the back burner and look out for frequent calls to similar equality checks.